### PR TITLE
Don't verify ECN counts for out of order ACKs

### LINKED ISSUE #2198: More ECN verification failure modes
(From #2189)

There's a perverse situation in which ECN verification results in another problem.

That situation is where the largest acknowledged is lower on packet that is sent later. AND that packet acknowledges more packets so that the counts are higher. AND that packet is reordered ahead of packets with higher largest acknowledged. This results in the verification failing on the second packet to arrive (counts are lower because it's an older ACK, but the largest acknowledged is higher).

A fix is to never decrease the largest acknowledged when sending an ACK frame. Another is to disable validation if the packet containing the ACK is not higher-numbered than the last packet containing an ACK. The question is whether we care enough to recommend either fix...

@gloinul, WDYT?

### PR DESCRIPTION
**Pull request by ianswett**
Fixes #2198

### CODE REVIEW

**Comment by ianswett** [comment]
There's probably slightly better phrasing for the packet number clause, so feel free to suggest something.

**Comment by ianswett** [comment]
Jana's talked me through this and now I realize that this is broken in a larger sense.  As we've previously discussed, any time an acknowledgement is dropped, ECN validation MAY fail.

**Comment by janaiyengar** [comment]
After more discussion, I agree that this does fix the problem. I think the text could be moved to a better location, but I'm fine with this for now.

**Comment by mikkelfj** [comment]
"Processed in different order than it was sent" is a bit vague. If there is reordering you don't know which frame is the one that is out of order. There needs to be a clear selection criteria such as not processing a frame in packet with a PN lower than a previously processed.

**Comment by janaiyengar** [comment]
@ianswett: I've reworded and moved the text. See what you think.

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -3065,6 +3064,10 @@ to be greater than the number of packets acknowledged in an ACK frame.  When
 this happens, and if verification succeeds, the local reference counts MUST be
 increased to match the counts in the ACK frame.
 
+Processing counts out of order can result in verification failure.  An endpoint
+MUST NOT perform this verification if the ACK frame is received in a packet with
+packet number lower than a previously received ACK frame.
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: With the proposed change, add:

> Verifying based on ACK frames that arrive out of order can result in disabling ECN unnecessarily.

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -3065,6 +3064,10 @@ to be greater than the number of packets acknowledged in an ACK frame.  When
 this happens, and if verification succeeds, the local reference counts MUST be
 increased to match the counts in the ACK frame.
 
+Processing counts out of order can result in verification failure.  An endpoint
+MUST NOT perform this verification if the ACK frame is received in a packet with
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]:
```suggestion
SHOULD NOT perform this verification if the ACK frame is received in a packet with
```

**Comment by janaiyengar** [comment]
@martinthomson : I've incorporated your comments.

**Comment by ianswett** [comment]
This still LGTM.

**Comment by martinthomson** [comment]
Thanks for considering my input.

**Pull request merged**